packages/diagram/src/layout/engine/layout.ts

```diff
@@ -0,0 +1,426 @@
+     *
+     * @param style defines all required layout information
+     * @returns the computed layout information
+     */
+    private computeLayoutInformation(style: { [key: string]: any }): LayoutInformation {
+        return {
+            marginTop: style.marginTop ?? style.margin ?? 0,
+            marginBottom: style.marginBottom ?? style.margin ?? 0,
+            marginLeft: style.marginLeft ?? style.margin ?? 0,
+            marginRight: style.marginRight ?? style.margin ?? 0
+        };
+    }
+
+    /**
+     * Calls measure on the element and generates a LayoutElement for it.
+     * computes the styles
+     *
+     * @param element the element to measure
+     * @param parent the parent element if existing
+     * @param constraints size constraitns required for measure
+     * @returns the generated LayoutElement
+     */
+    measure(element: FullObject, parent: LayoutElement | undefined, constraints: SizeConstraints): LayoutElement {
+        const type = assertString(element.getLocalFieldOrUndefined("type")!.value, "type");
+        const cls = nativeToList(element.getLocalFieldOrUndefined("class")?.value?.toNative() ?? {});
+        const id = this.generateId(parent);
+        const layoutElement: LayoutElement = {
+            id,
+            element,
+            parent,
+            styles: {},
+            layoutConfig: this.engine.layoutConfigs.get(type)!,
+            class: new Set(cls),
+            edits: generateEdits(element)
+        };
+        this.elementIdLookup.set(element, id);
+        this.layoutElementLookup.set(id, layoutElement);
+        this.applyStyles(layoutElement);
+        const styles = layoutElement.styles;
+        const layoutInformation = this.computeLayoutInformation(layoutElement.styles);
+        layoutElement.layoutInformation = layoutInformation;
+        const marginX = layoutInformation.marginLeft + layoutInformation.marginRight;
+        const marginY = layoutInformation.marginTop + layoutInformation.marginBottom;
+        const computedConstraints = this.computeSizeConstraints(styles, constraints, marginX, marginY);
+        const requestedSize = layoutElement.layoutConfig.measure(this, layoutElement, computedConstraints);
+        const computedSize = addToSize(requestedSize, marginX, marginY);
+        const realSize = matchToConstraints(computedSize, constraints);
+        layoutElement.measuredSize = realSize;
+        layoutElement.requestedSize = requestedSize;
+        return layoutElement;
+    }
+
+    /**
+     * Generates a new id based on the parent element
+     *
+     * @param parent the parent element
+     * @returns the generated id
+     */
+    private generateId(parent: LayoutElement | undefined): string {
+        const parentId = parent?.id ?? "";
+        const counter = this.elementIdCounter.get(parentId) ?? 0;
+        const id = `${parentId}_${counter}`;
+        this.elementIdCounter.set(parentId, counter + 1);
+        return id;
+    }
```

Comment:
Is Hylimo single or multithreaded?

Reply:
depends on how you use it
in general, the language server (and thus also all the stuff in diagram) runs on a separate thread, (there might be even multiple language servers which are connected in a main worker architecture)
so there is at least one thread, but for a reasonable editing experience, 3 are used by default

Reply:
The question is in particular if two threads can access the ID counter simultaneously.
Because if it is, you have a data race here.

Reply:
no, that's not possible
different layout engines can be used in different threads, but the same layout engine cannot be used in multiple threads (I don't even think that is possible in a browser environment without manually implementing a RPC)